### PR TITLE
[ENH] circvar calculated simply as 1-R

### PR DESCRIPTION
**Pull request by kleskjr**
The circular statistic function circvar is updated to estimate circular
variance simply as one minus the mean resultant vector (1-R), instead
of using approximation to the linear variance. Thus, the returned value
is in the range [0, 1]. This definition is compatible with the
literature (e.g., Fisher, 1993; Jammalamadaka and SenGupta, 2001) and
other implementations in statistical packages (e.g., R and Matlab).

### CODE REVIEW

File: scipy/stats/morestats.py [hunk 1]
```diff
@@ -2723,15 +2724,20 @@ def circvar(samples, high=2*pi, low=0, axis=None):
...
     Notes
     -----
-    This uses a definition of circular variance that in the limit of small
-    angles returns a number close to the 'linear' variance.
+    This uses the following definition of circular variance: ``1-R'', where 
+    ``R'' is the mean resultant vector. The returned value is in the range
+    [0, 1], 0 standing for no variance, and 1 for a large variance.
+
+    References
+    -----
```
Comment (larsmans) [inline scipy/stats/morestats.py]: "Title underline too short" on this line, and one more below.

**Comment by larsmans** [comment]
This is not backwards compatible, right? Wouldn't adding an option be a better idea?

**Comment by kleskjr** [comment]
No, it's not compatible. The `circvar` method used until now, takes simply `circstd` squared (before the normalization). However, I haven't seen yet such a definition anywhere else. The sources I've looked in are:
* [wikipedia] (https://en.wikipedia.org/wiki/Directional_statistics)
* [Matlab's Circular Statistics Toolbox] (http://www.mathworks.com/matlabcentral/fileexchange/10676-circular-statistics-toolbox--directional-statistics-/content/circ_var.m)
* [R's Circular Statistics Toolbox] (http://rpackages.ianhowson.com/rforge/circular/man/var.circular.html)
* Fisher NI, Statistical analysis of circular data. Cambridge, 1993.
* Jammalamadaka SR and SenGupta A, Topics in circular statistics, Singapore, 2001.

All of them define the circular variance as `1-R`. Only the last reference defines it as `n(1-R)`, where `n` is the array size. Therefore, the `circvar` definition used until now might not be correct.

Is a compatibility option still needed in this case?

**Comment by mirca** [comment]
How compatible (conceptually) ``circvar`` and ``circstd`` are going to be then?

**Comment by josef-pkt** [comment]
I don't remember any details for how transformation of circular data works.
What looks strange to me in this change is that the variance becomes independent of the units. This is usually not the case with any other random variables. In the case of 360 degrees, it looks strange having a tiny variance, as in the changed test case.

If this is just a renormalization of the variance into radians, then this should be an option not a backwards incompatible redefinition.

**Comment by kleskjr** [comment]
mirca said:

> How compatible (conceptually) `circvar` and `circstd` are going to be then?

`circstd` and `circvar` are not going to be conceptually compatible. Fisher, 1993 is explicit that circular `std` is not defined as circular `var**0.5`. 

josef-pkt said:
> What looks strange to me in this change is that the variance becomes independent of the units. This is usually not the case with any other random variables.

This is how circular variance is defined in the literature, with values between 0 and 1, and independent on the units.

> If this is just a renormalization of the variance into radians, then this should be an option not a backwards incompatible redefinition.

It's not only a renormalization. After normalisation, the PR's `circvar` is approximately twice smaller than the linear variance of data with very small deviations around the mean. For big variances, however, I would expect the old and new definition to deviate even after normalisation.

I would strongly suggest to keep the new definition of `circvar`. However, an optional variable could set whether we expect `1-R` as output or the de-normalised value `((high-low)/(2*pi))**2*(1-R)`. The same optional variable should also be implemented for the (de)normalisation of `cirstd`.

**Comment by josef-pkt** [comment]
3 years ago I was working on a circular "toolbox" based initially on a matlab translation.

I used var = 1 - R, and a wrapper that converts between the circle units (degrees, hours or whatever) into radians and back.
I don't remember much of the details but was reading mainly Mardia's book for this.

AFAIR (!), I never managed to figure out these scipy stats function and left them to Travis. (I thought differences might be due to scaling. There are also weird things in R packages, where I wasn't quite sure where their scaling or transformation comes from.)

**Comment by kleskjr** [comment]
A new push:

- added an optional parameter for result normalization to ``circstd`` and ``circvar``. Because of compatibility, the returned values are scaled by default depending on the variable range as it was until now.
- docstrins of ``circmean``, ``circstd`` and ``circvar`` have a more accurate description of the ``high`` and ``low`` parameters.

**Comment by ev-br** [comment]
I gather there might be some use of / experience with these sorts of functions in astropy land.

**Comment by mirca** [comment]
@ev-br You are right. See [#4472](https://github.com/astropy/astropy/pull/4472).

**Comment by zym1010** [comment]
+1 for this feature. Also, it might be also good to allow a `weights` argument.

**Comment by mdhaber** [comment]
@kleskjr I'd like to help you finish this up. Can you [select "Allow edits from maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)?

**Comment by mdhaber** [comment]
@kleskjr Thought I'd check in again. If you could allow [select "Allow edits from maintainers"](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork), I'd appreciate it!

**Comment by kleskjr** [comment]
@mdhaber, edits are allowed. Sorry for the delay!

**Comment by mdhaber** [comment]
Thanks @kleskjr. How important do you think the `normalize` option is? 

In `cirstd`, I think I see how having `low` and `high` could be useful without `normalize`:
* whatever your units were originally, you can use `low` and `high` to convert them to the range 0 to 2 pi. 
* Since the output can be interpreted as being in units of radians, `normalize` gives you the option to convert back to the original units. This is supported by the literature - at least Mardia (see gh-15652).

The validity of `normalize` for `circvar` is less clear to me. The output is not in radians; it is in the range of 0 to 1, and we can think of it as the distance from the outside of a unit circle, right? If that's correct, it seems that the more appropriate normalization would be in terms of the radius of the original circle, but we don't have that information.

File: scipy/stats/_morestats.py [hunk 2]
```diff
@@ -3590,8 +3595,16 @@ def circvar(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
...
+    for a large variance. In the limit of small angles, this value is similar
+    to half the 'linear' variance.
+
+    References
+    ----------
+    ..[1] Fisher, N.I. *Statistical analysis of circular data*. Cambridge
+          University Press, 1993.
 
     Examples
     --------
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:3610]: Failure is due to:
```
scipy.stats.circvar
-------------------

File "build/testenv/lib/python3.9/site-packages/scipy/stats/_morestats.py", line 3612, in circvar
Failed example:
    circvar([0, 2*np.pi/3, 5*np.pi/3])
Expected:
    2.19722457734
Got:
    0.6666666666666665
```
below.

**Comment by steppi** [comment]
> @steppi I suspect that the reason this hasn't already been fixed is that others are unsure of the math. That's why I tagged you, but NP if it's not of interest.

Definitely of interest but my time is scarce now. I think I can take a look in 2-3 weeks.

**Comment by rkern** [comment]
> I don't see support for this sort of thing in the literature, so I think it might be easier to merge this without `normalize`, at least in `circvar`.

@mdhaber I concur.

File: scipy/stats/_morestats.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_morestats.py:3610]: The new result is correct.
```
library(circular)
options(digits=16)
x = c(0, 2*pi/3, 5*pi/3)
var.circular(x)
```
Produces:
```

Attaching package: 'circular'

The following objects are masked from 'package:stats':

    sd, var

[1] 0.6666666666666665
```

File: scipy/stats/_morestats.py [hunk 3]
```diff
@@ -3643,17 +3653,20 @@ def circstd(samples, high=2*pi, low=0, axis=None, nan_policy='propagate'):
...
+        Axis along which standard deviations are computed. The default is
         to compute the standard deviation of the flattened array.
     nan_policy : {'propagate', 'raise', 'omit'}, optional
         Defines how to handle when input contains nan. 'propagate' returns nan,
         'raise' throws an error, 'omit' performs the calculations ignoring nan
         values. Default is 'propagate'.
+    normalize : boolean, optional
+        If True, the returned value is equal to ``sqrt(-2*log(R))`` and does
+        not depend on the variable units. If False (default), the returned
+        value is scaled by ``((high-low)/(2*pi))``.
```
Comment (mdhaber) [inline scipy/stats/_morestats.py:3669]:
```suggestion
    normalize : boolean, optional
        If True, the returned value is equal to ``sqrt(-2*log(R))`` as 
        described in the Notes; the output can be interpreted in units of 
        radians. If False (default), the returned value is scaled by 
        ``((high-low)/(2*pi))``.
```

**Comment by mdhaber** [comment]
Thanks @rkern. Other than the adjustment above, this is ready from my perspective. If you approve the PR and there are no concerns from the mailing list, I'll go ahead and merge it in a week.

**Review by steppi — APPROVED** [review]
This looks good to me. It's been 5 years since I've done anything with directional statistics, but I've double-checked @kleskjr's references and agree that this is the most reasonable definition of the circular variance. I also agree with not including a normalize option for `cirvar`.

**Comment by mdhaber** [comment]
There were no replies to the [mailing list post](https://mail.python.org/archives/list/scipy-dev@python.org/thread/UMVBWTZ2DHYNBQIYOAWQMKS2DUDRUAG5/#UMVBWTZ2DHYNBQIYOAWQMKS2DUDRUAG5), there is +1 explicit approval from @steppi (and me), and there seems to be other support above, so I'm merging. I added a note about the backward-incompatible change in the release notes.

**Pull request merged**